### PR TITLE
WT-7420 Tiered local storage changed to flush files to bucket directory.

### PR DESCRIPTION
**Pull request by ddanderson**
A new "cache_directory" config parameter is now available when
customizing a file system. New files are placed there, and copied
to the bucket directory only when flushed.  Currently no files
are "aged out" of cache nor are "paged in" to cache.

### CODE REVIEW

**Review by keitharnoldsmith — COMMENTED** [review]
Generally looks good.  But one question: Should we `fsync()` the bucket file after copying?  You're code doesn't do this, but in theory the caller could immediately delete the cache file once the copy is done and we wouldn't want to crash and have neither.

File: ext/storage_sources/local_store/local_store.c [hunk 1]
```diff
@@ -100,7 +100,8 @@ typedef struct {
     WT_FILE_SYSTEM *wt_fs;
 
     char *auth_token; /* Identifier for key management system */
-    char *bucket;     /* Actually a directory path for local implementation */
+    char *cache_dir;  /* Directory for pre-flushed objects and cached objects */
+    char *bucket_dir; /* Directory that stands in for cloud storage bucket */
```
Comment (sueloverso) [inline ext/storage_sources/local_store/local_store.c:104]: Alphabetize

File: ext/storage_sources/local_store/local_store.c [hunk 2]
```diff
@@ -304,6 +307,34 @@ local_flush_free(LOCAL_FLUSH_ITEM *flush)
...
+    struct stat sb;
+    int ret;
+    char *dirname;
+
+    if (len == -1)
+        len = (ssize_t)strlen(s);
+    dirname = strndup(s, (size_t)len + 1); /* Room for null */
+    if (dirname == NULL)
+        return (ENOMEM);
+    ret = stat(dirname, &sb);
```
Comment (sueloverso) [inline ext/storage_sources/local_store/local_store.c:326]: Do we need to make sure that any python test or other Csuite test using this is not Windows?

File: ext/storage_sources/local_store/local_store.c [hunk 3]
```diff
@@ -541,28 +599,79 @@ local_flush(WT_STORAGE_SOURCE *storage_source, WT_SESSION *session, WT_FILE_SYST
 static int
 local_flush_one(LOCAL_STORAGE *local, WT_SESSION *session, LOCAL_FLUSH_ITEM *flush)
 {
-    int ret;
+    WT_FILE_SYSTEM *wt_fs;
+    WT_FILE_HANDLE *dest, *src;
```
Comment (sueloverso) [inline ext/storage_sources/local_store/local_store.c:603]: Alphabetize `WT_FILE_HANDLE` before `WT_FILE_SYSTEM`

**Review by sueloverso — APPROVED** [review]
LGTM with a few minor questions.

File: ext/storage_sources/local_store/local_store.c (hunk 2 above)
Comment (ddanderson) [reply ext/storage_sources/local_store/local_store.c:326]: The ext/storage_source/local_store/ directory is configured only to be built on POSIX.  The only test that uses local_store right now, "skips" it if local_store is not built, we'll need to do that for any new tests that use local_store - it's easy to do.  I suspect if we have any Csuite tests in the future (like abort_flush_tier), we'll enable that for only POSIX.  So I think we're covered now.

**Comment by ddanderson** [comment]
@sueloverso and @keitharnoldsmith, since Windows came up in the review, I figured I'd share some thoughts I have about the question: should the storage extensions support Windows support in the future?

I think we'd run into some practical issues, especially as we evolve toward "drivers" for real cloud providers.  We'll need replacements for any direct system calls of course (`stat` came up in this review), and I suspect we'll eventually need a pool of worker threads to manage asynchronous flushing of multiple objects, so some abstraction for threads.  WT has all the abstractions for these, but they aren't easily available to an extension.  And I don't see any pressing need to support Windows for these "drivers", so I would propose we "ignore" it until we think there's a need.

**Comment by sueloverso** [comment]
I am happy to ignore Windows until we need something.

**Comment by ddanderson** [comment]
> Generally looks good. But one question: Should we `fsync()` the bucket file after copying? You're code doesn't do this, but in theory the caller could immediately delete the cache file once the copy is done and we wouldn't want to crash and have neither.

Sounds good, I made this change.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM

**Pull request merged**